Sources/Build/BuildPlan.swift

```diff
@@ -657,24 +685,60 @@ public final class SwiftTargetBuildDescription {
         self.tempsPath = buildParameters.buildPath.appending(component: target.c99name + ".build")
         self.derivedSources = Sources(paths: [], root: tempsPath.appending(component: "DerivedSources"))
         self.pluginDerivedSources = Sources(paths: [], root: buildParameters.dataPath)
+        self.pluginDerivedResources = []
         self.buildToolPluginInvocationResults = buildToolPluginInvocationResults
         self.prebuildCommandResults = prebuildCommandResults
+        self.observabilityScope = observabilityScope
+
+        func handlePluginDerivedFile(path absPath: AbsolutePath) {
+            // 5.6 handled treated all generated files as sources.
+            if toolsVersion <= .v5_6 {
+                let relPath = absPath.relative(to: self.pluginDerivedSources.root)
+                self.pluginDerivedSources.relativePaths.append(relPath)
+                return
+            }
+
+            var rule = TargetSourcesBuilder.computeRule(for: absPath, toolsVersion: toolsVersion, additionalFileRules: additionalFileRules, observabilityScope: observabilityScope)
+
+            // If we did not find a rule for a generated file, we treat it as to be processed for now. Eventually, we should handle generated files the same as other files and require explicit handling in the manifest for unknown types.
+            if rule == .none {
+                rule = .processResource(localization: .none)
+            }
+
+            switch rule {
+            case .compile:
+                if absPath.extension == "swift" {
+                    let relPath = absPath.relative(to: self.pluginDerivedSources.root)
+                    self.pluginDerivedSources.relativePaths.append(relPath)
+                } else {
+                    observabilityScope.emit(error: "Generated plugin source files can only use Swift: \(absPath)")
+                }
+            case .copy, .processResource:
+                if let resource = TargetSourcesBuilder.resource(for: absPath, with: rule, defaultLocalization: target.defaultLocalization, targetName: target.name, targetPath: target.underlyingTarget.path, observabilityScope: observabilityScope) {
+                    self.pluginDerivedResources.append(resource)
+                } else {
+                    // If this is reached, `TargetSourcesBuilder` already emitted a diagnostic, so we can ignore this case here.
+                }
+            case .header:
+                observabilityScope.emit(error: "Headers generated by plugins are not supported at this time: \(absPath)")
```

Comment:
This is good as a start — a future PR could probably consolidate these into a single error with an associated list of headers, but can definitely be a future PR.